uccb/src/bin/validator-oracle.rs

```diff
@@ -0,0 +1,273 @@
+                    )
+                    .await
+                    .is_err()
+                    {
+                        error!(
+                            "Failed updating the validator manager on {}",
+                            chain_client.rpc_url
+                        );
+                    }
+
+                    if receiver.changed().await.is_err() {
+                        break;
+                    }
+                }
+            });
+            handles.push(handle);
+        }
+
+        let result = self.listen_to_staker_updates(sender).await;
+
+        for handle in handles {
+            handle.await?;
+        }
+
+        result
+    }
+
+    async fn listen_to_staker_updates(
+        &mut self,
+        sender: watch::Sender<Vec<Address>>,
+    ) -> Result<()> {
+        let filter = Filter::new()
+            .address(contract_addr::DEPOSIT)
+            // Must be the same signature as the event in deposit.sol
+            .event("StakerAdded(bytes)")
+            .from_block(BlockNumberOrTag::Finalized);
+
+        let subscription = self
+            .zq2_chain_client()
+            .provider
+            .subscribe_logs(&filter)
+            .await?;
+        let mut stream = subscription.into_stream();
+        while let Some(log) = stream.next().await {
+            // TODO: infer if staker added or removed
+            info!("Received validator update: {log:?}");
+
+            let validators = self.get_stakers().await?;
+            info!(
+                "Updating chains to the current validator set to: {}",
+                Display(&validators)
+            );
+
+            sender.send(validators)?;
+        }
+
+        Ok(())
+    }
+
+    async fn create_zq2_chain_client(
+        config: Config,
+        signer: &PrivateKeySigner,
+    ) -> Result<ChainClient> {
+        ChainClient::new(
+            &ChainConfig {
+                rpc_url: config.zq2.rpc_url,
+                chain_gateway_address: config.zq2.chain_gateway_address,
+                chain_gateway_block_deployed: 0,
+                block_instant_finality: false,
+                legacy_gas_estimation: false,
+            },
+            config.zq2.validator_manager_address,
+            signer.clone(),
+        )
+        .await
+    }
+
+    fn zq2_chain_client(&self) -> &ChainClient {
+        &self.chain_clients[0]
+    }
+
+    async fn get_stakers(&self) -> Result<Vec<Address>> {
+        debug!("Retreiving validators from the deposit contract");
+
+        let contract: ContractInstance<PubSubFrontend, _> = ContractInstance::new(
+            contract_addr::DEPOSIT,
+            self.zq2_chain_client().provider.as_ref(),
+            Interface::new(self.deploy_abi.clone()),
+        );
+
+        let call_builder: DynCallBuilder<_, _, _> =
+            contract.function("getStakerSignerAddresses", &[])?;
+        let output = call_builder.call().await?;
+        let validators = if output.len() == 1 {
```

Comment:
The error handling in this method is inconsistent IMO

* If the output doesn't have length 1, we silently return an empty vector - Is this expected to happen?
* If the call fails, we return an `Err`.
* If the output has length one but isn't of the expected format, we panic (via `.unwrap()`).

Conceptually I think all of these should be `Err`s instead?

Reply:
Well, not exactly. Some of this code could be spared if there was an easy way to use statically-typed contracts (like when using abigen!). Didn't seem to work with sol! last time. I suppose we could panic in each case since in order for the oracle to function properly, no error should be expected really.

Reply:
Sure - But if our ABI is wrong or we're pointing at the wrong contract then these panics will occur. I think it makes sense to return them as errors instead.

Reply:
We are still panicking here.